### PR TITLE
[operator] Enable Audit Logs in Guest Clusters

### PR DESCRIPTION
**Pull request by rschalo**
### Description of changes:
This PR adds audit logging to guest clusters by default.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rschalo** [comment]
Still need to show the logs in CloudWatch - this is not ready.

File: operator/pkg/controllers/master/auditlog.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"context"
+	"fmt"
+
+	"github.com/awslabs/kubernetes-iteration-toolkit/operator/pkg/apis/controlplane/v1alpha1"
+	"github.com/awslabs/kubernetes-iteration-toolkit/operator/pkg/utils/object"
+	v1 "k8s.io/api/core/v1"
+)
+
+func (c *Controller) reconcileAuditLogConfig(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	providerConfig := auditLogConfig
```
Comment (prateekgogia) [inline operator/pkg/controllers/master/auditlog.go:27]: Does `GenerateConfigMap` mutate the config arg passed? If not, can we directly use auditLogConfig?

File: operator/pkg/controllers/master/kubeapiserver.go [hunk 2]
```diff
@@ -141,6 +141,9 @@ func apiServerPodSpecFor(controlPlane *v1alpha1.ControlPlane) v1.PodSpec {
 					"--tls-private-key-file=/etc/kubernetes/pki/apiserver/apiserver.key",
 					"--authentication-token-webhook-config-file=/var/aws-iam-authenticator/kubeconfig/kubeconfig.yaml",
 					"--encryption-provider-config=/etc/kubernetes/aws-encryption-provider/encryption-configuration.yaml",
+					"--audit-policy-file=/etc/kubernetes/audit-policy/audit-policy.yaml",
+					"--audit-log-path=/var/log/kubernetes/audit/audit.log",
+					"--audit-log-maxbackup=1",
```
Comment (prateekgogia) [inline operator/pkg/controllers/master/kubeapiserver.go:146]: Do we know whats the max log file size before it rotates? If its too big, we can use this flag as well `--audit-log-maxsize`

File: operator/pkg/controllers/master/kubeapiserver.go [hunk 3]
```diff
@@ -389,6 +400,21 @@ func apiServerPodSpecFor(controlPlane *v1alpha1.ControlPlane) v1.PodSpec {
...
+		}, {
+			Name: "audit-log",
+			VolumeSource: v1.VolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Path: "/var/log/kubernetes/audit/",
+					Type: &hostPathDirectoryOrCreate,
+				},
+			},
+		}, {
+			Name: "audit",
```
Comment (prateekgogia) [inline operator/pkg/controllers/master/kubeapiserver.go:412]: nit: `audit`-> `audit-config`

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go [hunk 4]
```diff
@@ -288,6 +313,9 @@ spec:
         - name: dmesg
           mountPath: /var/log/dmesg
           readOnly: true
+        - name: audit-log
+          mountPath: /var/log/kubernetes/audit/
+          readOnly: false
```
Comment (prateekgogia) [inline substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:318]: We already mount `/var/log` on line 302, do we need to mount audit directory again?

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go [hunk 5]
```diff
@@ -333,12 +364,12 @@ func (t *FluentBit) Create(ctx context.Context, substrate *v1alpha1.Substrate) (
 		return reconcile.Result{}, fmt.Errorf("initializing client, %w", err)
 	}
 
-  	// Apply fluentbit ns
-	  if err := client.ApplyYAML(ctx, []byte(fluentbitNS)); err != nil {
+	// Apply fluentbit ns
+	if err := client.ApplyYAML(ctx, []byte(fluentbitNS)); err != nil {
```
Comment (prateekgogia) [inline substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:368]: Thanks for fixing these!

**Review by prateekgogia — COMMENTED** [review]
Looks good overall, left some feedback. I am going to try the audit log config in my cluster tomorrow.

File: operator/pkg/controllers/master/kubeapiserver.go (hunk 2 above)
Comment (rschalo) [reply operator/pkg/controllers/master/kubeapiserver.go:146]: I believe max log size is 100MB by default. Did we want to decrease this?

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go (hunk 4 above)
Comment (rschalo) [reply substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:318]: The audit log volume cannot be `readOnly: true`. Did we want to update the `/var/log` volume mount to be `readOnly: false`?

File: operator/pkg/controllers/master/auditlog.go (hunk 1 above)
Comment (rschalo) [reply operator/pkg/controllers/master/auditlog.go:27]: Looks like it returns a ConfigMap and does not mutate the passed string. Updated to use `auditLogConfig` directly.

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go (hunk 4 above)
Comment (prateekgogia) [reply substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:318]: Why does Fluentbit needs to write to audit logs file? It should be similar to any other logs file which FluentBit is parsing and pushing to CW

Comment (rschalo) [reply substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go]: Removed the extra mount and tested. Still populating `audit.log` as expected.

File: operator/pkg/controllers/master/kubeapiserver.go [hunk 6]
```diff
@@ -141,6 +141,9 @@ func apiServerPodSpecFor(controlPlane *v1alpha1.ControlPlane) v1.PodSpec {
 					"--tls-private-key-file=/etc/kubernetes/pki/apiserver/apiserver.key",
 					"--authentication-token-webhook-config-file=/var/aws-iam-authenticator/kubeconfig/kubeconfig.yaml",
 					"--encryption-provider-config=/etc/kubernetes/aws-encryption-provider/encryption-configuration.yaml",
+					"--audit-policy-file=/etc/kubernetes/audit-policy/audit-policy.yaml",
+					"--audit-log-path=/var/log/kubernetes/audit/audit.log",
```
[the remark this answers is not in the document]
Comment (prateekgogia) [reply operator/pkg/controllers/master/kubeapiserver.go:145]: You will need to include the POD_NAME env -
```
{
	Name: "POD_NAME",
	ValueFrom: &v1.EnvVarSource{
		FieldRef: &v1.ObjectFieldSelector{
		       FieldPath: "metadata.name",
		},
	},
}
```

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go [hunk 7]
```diff
@@ -152,7 +152,22 @@ data:
         Refresh_Interval    10
         Read_from_Head      ${READ_FROM_HEAD}
         Path_Key            path
- 
+
+    [INPUT]
+        Name                tail
+        Tag                 kube_audit.*
+        Path                /var/log/kubernetes/audit/audit.log
```
Comment (prateekgogia) [inline substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:159]:
```suggestion
        Tag                 audit.*
        Path                /var/log/kubernetes/audit/*.log
```

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go [hunk 8]
```diff
@@ -174,7 +189,17 @@ data:
         auto_create_group   true
         extra_user_agent    container-insights
         log_retention_days  30
- 
+
+    [OUTPUT]
+        Name                cloudwatch
+        Match               kube_audit.*
```
Comment (prateekgogia) [inline substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:195]:
```suggestion
        Match               audit.*
```

File: substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go [hunk 9]
```diff
@@ -174,7 +189,17 @@ data:
...
+        region              ${AWS_REGION}
+        log_group_name      /${CLUSTER_NAME}
+        log_stream_name     audit-log
```
Comment (prateekgogia) [inline substrate/pkg/controller/substrate/cluster/addons/fluentbitcw.go:198]:
```suggestion
        log_stream_prefix   apiserver-
```

**Review by prateekgogia — APPROVED** [review]
LGTM! Great work!

**Pull request merged**